File: api/v1/schemas/recommendation_tracking.py
# -*- coding: utf-8 -*-
"""历史推荐追踪 API Schema — 请求/响应模型."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field, ConfigDict


# ── 请求模型 ────────────────────────────────────────────────────────────────


class CreateRecommendationRequest(BaseModel):
    """创建推荐记录请求."""
    code: str = Field(..., description="股票代码", examples=["000001"])
    trade_date: str = Field(..., description="交易日期 (YYYY-MM-DD)", examples=["2026-06-13"])
    recommendation_price: float = Field(..., description="推荐价格", examples=[12.50])
    signal: str = Field(default="buy", description="推荐方向: buy/sell/hold")
    source: str = Field(default="analysis", description="来源: analysis/deep_analysis/seal_plate/comprehensive/manual")
    source_task_id: Optional[str] = Field(None, description="来源任务ID")
    reason: Optional[str] = Field(None, description="推荐理由")

    model_config = ConfigDict(json_schema_extra={
        "example": {
            "code": "000001",
            "trade_date": "2026-06-13",
            "recommendation_price": 12.50,
            "signal": "buy",
            "source": "deep_analysis",
            "source_task_id": "abc123def456",
            "reason": "多维度分析看多信号，技术面+基本面共振"
        }
    })


class UpdateRecordRequest(BaseModel):
    """更新推荐记录请求（部分更新）."""
    code: Optional[str] = None
    trade_date: Optional[str] = None
    signal: Optional[str] = None
    recommendation_price: Optional[float] = None
    current_price: Optional[float] = None
    status: Optional[str] = None
    reason: Optional[str] = None
    notes: Optional[str] = None


class CloseRecordRequest(BaseModel):
    """平仓请求."""
    close_price: float = Field(..., description="平仓价格")

    model_config = ConfigDict(json_schema_extra={
        "example": {"close_price": 13.80}
    })


class UpdatePriceRequest(BaseModel):
    """更新当前价格请求."""
    current_price: float = Field(..., description="当前价格")

    model_config = ConfigDict(json_schema_extra={
        "example": {"current_price": 13.15}
    })


class SummaryRequest(BaseModel):
    """生成自省总结请求."""
    start_date: Optional[str] = Field(None, description="开始日期 (YYYY-MM-DD)")
    end_date: Optional[str] = Field(None, description="结束日期 (YYYY-MM-DD)")

    model_config = ConfigDict(json_schema_extra={
        "example": {"start_date": "2026-05-01", "end_date": "2026-06-13"}
    })


# ── 响应模型 ────────────────────────────────────────────────────────────────


class RecommendationRecordSchema(BaseModel):
    """推荐记录响应."""
    id: int
    code: str
    trade_date: str
    recommendation_time: str
    signal: str
    recommendation_price: float
    current_price: Optional[float]
    price_deviation_pct: Optional[float]
    status: str
    close_price: Optional[float]
    close_date: Optional[str]
    profit_loss_pct: Optional[float]
    source: str
    source_task_id: Optional[str]
    reason: str
    notes: str
    created_at: str
    updated_at: str


class RecommendationListResponse(BaseModel):
    """推荐记录列表响应."""
    total: int
    page: int
    limit: int
    items: List[RecommendationRecordSchema]


class SignalStat(BaseModel):
    """按方向的分项统计."""
    total: int
    wins: int
    win_rate: float
    avg_pl_pct: float


class SourceStat(BaseModel):
    """按来源的分项统计."""
    total: int
    wins: int
    win_rate: float
    avg_pl_pct: float


class ActiveDeviationItem(BaseModel):
    """活跃持仓偏差项."""
    id: int
    code: str
    trade_date: str
    signal: str
    recommendation_price: float
    current_price: float
    deviation_pct: float


class RecommendationStatsResponse(BaseModel):
    """推荐胜率与偏差统计响应."""
    total_records: int
    active_count: int
    closed_count: int
    win_rate: float
    win_count: int
    loss_count: int
    avg_pl_pct: float
    avg_win_pl_pct: float
    avg_loss_pl_pct: float
    max_win_pct: float
    max_loss_pct: float
    profit_factor: float
    total_pl_pct: float
    by_signal: Dict[str, SignalStat] = {}
    by_source: Dict[str, SourceStat] = {}
    active_deviation: List[ActiveDeviationItem] = []


class SummaryResponse(BaseModel):
    """自省总结响应."""
    stats: RecommendationStatsResponse
    summary: str

File: apps/dsa-web/src/api/recommendationTracking.ts
import apiClient from './index';
import { toCamelCase } from './utils';

export interface RecommendationRecord {
  id: number;
  code: string;
  tradeDate: string;
  recommendationTime: string;
  signal: 'buy' | 'sell' | 'hold';
  recommendationPrice: number;
  currentPrice: number | null;
  priceDeviationPct: number | null;
  status: 'active' | 'closed' | 'expired';
  closePrice: number | null;
  closeDate: string | null;
  profitLossPct: number | null;
  source: string;
  sourceTaskId: string | null;
  reason: string;
  notes: string;
  createdAt: string;
  updatedAt: string;
  // 交易信号增强字段
  signalType: string;
  strategyPattern: string;
  confidence: number;
  entryMethod: string;
  stopLoss: number | null;
  takeProfit: number | null;
  sectors: string;
  sentimentPhase: string;
  expectedHoldDays: number | null;
  timeHorizon: string;
}

export interface RecommendationListResponse {
  total: number;
  page: number;
  limit: number;
  items: RecommendationRecord[];
}

export interface SignalStat {
  total: number;
  wins: number;
  winRate: number;
  avgPlPct: number;
}

export interface SourceStat {
  total: number;
  wins: number;
  winRate: number;
  avgPlPct: number;
}

export interface ActiveDeviationItem {
  id: number;
  code: string;
  tradeDate: string;
  signal: string;
  recommendationPrice: number;
  currentPrice: number;
  deviationPct: number;
}

export interface RecommendationStats {
  totalRecords: number;
  activeCount: number;
  closedCount: number;
  winRate: number;
  winCount: number;
  lossCount: number;
  avgPlPct: number;
  avgWinPlPct: number;
  avgLossPlPct: number;
  maxWinPct: number;
  maxLossPct: number;
  profitFactor: number;
  totalPlPct: number;
  bySignal: Record<string, SignalStat>;
  bySource: Record<string, SourceStat>;
  activeDeviation: ActiveDeviationItem[];
}

export interface SummaryResponse {
  stats: RecommendationStats;
  summary: string;
}

export interface CreateRecordParams {
  code: string;
  tradeDate: string;
  recommendationPrice: number;
  signal?: string;
  source?: string;
  sourceTaskId?: string;
  reason?: string;
}

export interface ListRecordsParams {
  code?: string;
  status?: string;
  signal?: string;
  source?: string;
  startDate?: string;
  endDate?: string;
  page?: number;
  limit?: number;
  tagFilter?: string;
}

// ── 共同点分析 ──────────────────────────────────────────────────

export interface CommonalityTag {
  key: string;
  label: string;
  value: string;
  count: number;
  category: string;
}

export interface CommonalityGroup {
  category: string;
  categoryLabel: string;
  tags: CommonalityTag[];
}

export interface CommonalityResponse {
  totalAnalyzed: number;
  groups: CommonalityGroup[];
}

export const recommendationTrackingApi = {
  /** 创建推荐记录 */
  createRecord: (params: CreateRecordParams) =>
    apiClient.post<RecommendationRecord>('/api/v1/recommendation-tracking/records', params)
      .then(res => toCamelCase<RecommendationRecord>(res.data)),

  /** 查询列表 */
  listRecords: (params: ListRecordsParams = {}) =>
    apiClient.get<RecommendationListResponse>('/api/v1/recommendation-tracking/records', { params })
      .then(res => toCamelCase<RecommendationListResponse>(res.data)),

  /** 获取单条详情 */
  getRecord: (id: number) =>
    apiClient.get<RecommendationRecord>(`/api/v1/recommendation-tracking/records/${id}`)
      .then(res => toCamelCase<RecommendationRecord>(res.data)),

  /** 更新记录 */
  updateRecord: (id: number, updates: Record<string, unknown>) =>
    apiClient.put<RecommendationRecord>(`/api/v1/recommendation-tracking/records/${id}`, updates)
      .then(res => toCamelCase<RecommendationRecord>(res.data)),

  /** 删除记录 */
  deleteRecord: (id: number) =>
    apiClient.delete<{ message: string; recordId: string }>(`/api/v1/recommendation-tracking/records/${id}`)
      .then(res => res.data),

  /** 平仓 */
  closeRecord: (id: number, closePrice: number) =>
    apiClient.post<RecommendationRecord>(`/api/v1/recommendation-tracking/records/${id}/close`, { close_price: closePrice })
      .then(res => toCamelCase<RecommendationRecord>(res.data)),

  /** 更新当前价格 */
  updatePrice: (id: number, currentPrice: number) =>
    apiClient.post<RecommendationRecord>(`/api/v1/recommendation-tracking/records/${id}/price`, { current_price: currentPrice })
      .then(res => toCamelCase<RecommendationRecord>(res.data)),

  /** 获取统计数据 */
  getStats: (startDate?: string, endDate?: string) =>
    apiClient.get<RecommendationStats>('/api/v1/recommendation-tracking/stats', {
      params: { start_date: startDate, end_date: endDate },
    }).then(res => toCamelCase<RecommendationStats>(res.data)),

  /** 生成自省总结 */
  generateSummary: (startDate?: string, endDate?: string) =>
    apiClient.post<SummaryResponse>('/api/v1/recommendation-tracking/summary', {
      start_date: startDate,
      end_date: endDate,
    }).then(res => toCamelCase<SummaryResponse>(res.data)),

  /** 获取推荐共同点分析 */
  getCommonality: (startDate?: string, endDate?: string, tagFilter?: string) =>
    apiClient.get<CommonalityResponse>('/api/v1/recommendation-tracking/commonality', {
      params: { start_date: startDate, end_date: endDate, tag_filter: tagFilter },
    }).then(res => toCamelCase<CommonalityResponse>(res.data)),
};
